### PR TITLE
[NPU]adaptation to support deterministic inference

### PR DESCRIPTION
**Pull request by Estrella-xx**
## Motivation

We need to enable deterministic inference on the Ascend NPU to ensure consistent LLM outputs across runs when the --enable-deterministic-inference parameter is set, when graph mode is disabled.

## Modifications

1.Replace the NPU operators mm, matmul, sum, mean, _log_softmax, npu_fused_infer_attention_score, and npu_add_rms_norm with batch-invariant implementations to ensure deterministic inference.
2.During the sampling stage, replace torch.multinomial with a deterministic batched alternative.
3.Support enabling chunking and radix cache during the prefill stage.

## Accuracy Tests

1.Accuracy of gsm8k dataset

- temperature=0
8 gsm8k-temperature=0

- temperature=0.7
9 gsm8k-temperature=0 7

2.Test determinism with varying batch sizes

- temperature=0
1 single-temperature=0

- temperature=0.7
2 single-temperature=0 7-new

3.Test determinism across different batch sizes given shared prefixes.

- temperature=0
3 prefix-return_logprob-temperaature=0

- temperature=0.7
4 prefix-return_logprob-temperaature=0 7

4.Test radix cache determinism (cached vs uncached prefill)

- temperature=0
5 radix_cache-temperature=0

- temperature=0.7
6 radix_cache-temperature=0 7

5.Single curl request to call the OpenAI API

7 api-v1_chat

6.Send a single curl request to the /generate endpoint

- temperature=0
10 api-generate-temperature=0

- temperature=0, return logprob
11 api-generate-temperature=0-return_logprob

- temperature=1
12 api-generate-temperature=1

- temperature=1, return logprob
13 api-generate-temperature=1-return_logprob

- temperature=1, top_k=10
14 api-generate-temperature=1-top_k

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

---
### CI States

Latest PR Test (Base): :x: [Run #27124705087](https://github.com/sgl-project/sglang/actions/runs/27124705087)
Latest PR Test (Extra): :x: [Run #27124704787](https://github.com/sgl-project/sglang/actions/runs/27124704787)

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 1]
```diff
@@ -832,30 +832,35 @@ def forward_extend(
...
-                            v[None, q_len_offset : q_len_offset + q_len],
-                            num_heads=layer.tp_q_head_num,
-                            num_key_value_heads=layer.tp_k_head_num,
-                            input_layout="BSND",  # todo, TND not supports q_heads!=k_heads
-                            atten_mask=self.fia_mask.unsqueeze(0),
-                            sparse_mode=3 if q_len != 1 else 0,
-                            scale=layer.scaling,
-                            next_tokens=0,
-                        )[0]
+                actual_seq_lengths_kv = [
```
Comment (Todobe) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: Check if `self.forward_metadata.seq_lens_cpu_int` can replace it.

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py [hunk 2]
```diff
@@ -929,6 +938,21 @@ def rms_norm_batch_invariant(
     return rms_norm(input, weight, eps=eps)
 
 
+def add_rms_norm(
```
Comment (Todobe) [inline python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: Add an npu suffix to distinguish this operator from a common operator.

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py (hunk 1 above)
Comment (Estrella-xx) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: done

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py (hunk 2 above)
Comment (Estrella-xx) [reply python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: done

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py [hunk 3]
```diff
@@ -6,12 +6,16 @@
 from typing import Any, Dict
 
 import torch
+import torch_npu
```
Comment (Todobe) [inline python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: `torch_npu` should be moved under if `_is_npu` after line 17.

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py [hunk 4]
```diff
@@ -577,6 +581,11 @@ def _log_softmax_batch_invariant(input, dim, _half_to_float):
     return log_softmax(input, dim=dim)
 
 
+def _npu_log_softmax_batch_invariant(input, dim, _half_to_float):
+    assert not _half_to_float, "not implemented"
```
Comment (Todobe) [inline python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: Check if `torch.ops.batch_invariant_ops.npu_log_softmax_batch_invariant` support not _half_to_float.

File: python/sglang/srt/server_args.py [hunk 5]
```diff
@@ -3332,10 +3332,13 @@ def _handle_deterministic_inference(self):
                 self.enable_aiter_allreduce_fusion = False
 
             # Check sampling backend
-            self.sampling_backend = "pytorch"
-            logger.warning(
-                "Sampling backend is set to pytorch for deterministic inference."
-            )
+            if self.attention_backend == "ascend":
```
Comment (Todobe) [inline python/sglang/srt/server_args.py]: The logic should be changed to the following: If sampling_backend is not ascend, change it to pytorch. Otherwise, no modification is required.

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py (hunk 3 above)
Comment (Estrella-xx) [reply python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: done

File: python/sglang/srt/server_args.py (hunk 5 above)
Comment (Estrella-xx) [reply python/sglang/srt/server_args.py]: done

File: python/sglang/srt/mem_cache/hisparse_memory_pool.py [hunk 6]
```diff
@@ -104,6 +104,8 @@ def get_mla_kv_buffer(
         return super().get_mla_kv_buffer(layer, loc, dst_dtype)
 
     def transfer_values_on_device(self, dst_indices, src_indices):
+        from sgl_kernel.kvcacheio import transfer_kv_all_layer_mla
```
Comment (Todobe) [inline python/sglang/srt/mem_cache/hisparse_memory_pool.py]: This should be deleted

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py [hunk 7]
```diff
@@ -577,10 +582,21 @@ def _log_softmax_batch_invariant(input, dim, _half_to_float):
     return log_softmax(input, dim=dim)
 
 
+def _npu_log_softmax_batch_invariant(input, dim, _half_to_float):
```
Comment (Todobe) [inline python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: All functions about npu should move to hardware_backend/npu

File: python/sglang/srt/mem_cache/hisparse_memory_pool.py (hunk 6 above)
Comment (chx96642264) [reply python/sglang/srt/mem_cache/hisparse_memory_pool.py]: done

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py (hunk 7 above)
Comment (chx96642264) [reply python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: done

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py [hunk 8]
```diff
@@ -10,8 +10,25 @@
 import triton.language as tl
 
 from sglang.srt.layers.deep_gemm_wrapper.configurer import ENABLE_JIT_DEEPGEMM
+from sglang.srt.utils import is_npu
 from sglang.srt.utils.common import calc_diff, get_bool_env_var
 
+_is_npu = is_npu()
+if _is_npu:
+    import batch_invariant_ops  # noqa: F401
```
Comment (Todobe) [inline python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: move this import to npu_batch_invariant_ops.py

Comment (chx96642264) [reply python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: done

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 9]
```diff
@@ -938,30 +938,28 @@ def forward_extend(
                 return attn_out
 
             if self.use_fia:
-                """FIA will support multi-bs in the later version of CANN"""
                 q = q.reshape(-1, layer.tp_q_head_num, layer.qk_head_dim)
```
Comment (Hexq0210) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:1092]: There might be padding issues.

Comment (chx96642264) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:1092]: done

File: python/sglang/srt/hardware_backend/npu/batch_invariant_ops/npu_batch_invariant_ops.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    else:
+        assert input.dtype in {
+            torch.float16,
+            torch.bfloat16,
+            torch.float32,
+        }, "only float types supported for now"
+        n_elems = 1
+        for d in dim:
+            n_elems *= input.shape[d]
+        return torch.sum(input, dim=dim, keepdim=keepdim, dtype=torch.float32) / n_elems
```
Comment (Hexq0210) [inline python/sglang/srt/hardware_backend/npu/batch_invariant_ops/npu_batch_invariant_ops.py:33]: Is this operator batch-invariant?

Comment (chx96642264) [reply python/sglang/srt/hardware_backend/npu/batch_invariant_ops/npu_batch_invariant_ops.py:33]: What I understand is that those with `dim` will be replaced, while those without `dim` will not be replaced. This is because during scheduling, many operations such as length calculations are performed, and the `int` type does not have uncertainty, so I think there is no need to replace it. Previously, if all `sum` operations were replaced, these `sum` operations without `dim` would result in errors, so only those with `dim` were replaced. In our SGLang, we actually do not use the invariant sum in any place. 

The GPU also does not replace this operator at all.
